crates/red_knot_python_semantic/resources/mdtest/boundness_declaredness/public.md

```diff
@@ -161,36 +183,46 @@ seems inconsistent when compared to the case just above.
 def flag() -> bool: ...
 
 if flag():
-    x: int
+    a: int
 ```
 
 ```py
 # TODO: this should raise an error. Once we fix this, update the section description and the table
 # on top of this document.
-from mod import x
+from mod import a
 
-reveal_type(x)  # revealed: int
+reveal_type(a)  # revealed: int
 
-# External modifications to `x` that violate the declared type are not allowed:
+# External modifications to `a` that violate the declared type are not allowed:
 # error: [invalid-assignment]
-x = None
+a = None
 ```
 
 ## Undeclared
 
 ### Undeclared but bound
 
+If a symbols is undeclared, we use the union of `Unknown` with the inferred type. Note that we treat
+symbols that are undeclared differently from symbols that are explicitly declared as `Unknown`:
+
 ```py path=mod.py
-x = 1
+from knot_extensions import Unknown
+
+# Undeclared:
+a = 1
+
+# Declared with `Unknown`:
+b: Unknown = 1
```

Comment:
I think something that is noteworthy here is that the entire purpose of `Unknown` (as distinct from `Any`) is to distinguish a dynamic type that originates from an explicit annotation vs one that originates from lack of annotation. So the whole idea of "declared with `Unknown`" is a contradiction. We have made it _possible_ by introducing `knot_extensions.Unknown`, but perhaps that is just confusing and was a mistake? It should really only be used in tests, but I would guess that in most cases we could structure the test so as to generate an `Unknown` type via lack-of-annotation instead (e.g. an un-annotated function parameter).

Reply:
Actually I just realized that "declared with Unknown" is not a contradiction; it could occur anytime someone annotates with a name we can't resolve, or an invalid type expression.

Still not totally clear to me whether tests for this case are better written using `knot_extensions.Unknown` (which is clear and explicit, but not "realistic") or using an actual unknown name (which then requires suppressing the unknown-name diagnostic.) I can see good arguments both ways, so no strong feelings.